### ISSUE TITLE
How to run a maximum number of tasks concurrently?

### ISSUE THREAD

**Issue by rightaway** [tj/co#194]
If I want to run something sequentially in co, I can do as the docs say:

```
yield a();
yield b();
```

If I want to do it in parallel I can do:

```
var array = [a(), b()];
yield array;
```

But how can I implement running a maximum number of things at a time? For example, if I have a very large array of URLs that need to be retrieved, I'd only want a max of 5 running at a time. Using the above examples I could either retrieve them one by one or all together. I'm having trouble wrapping my head around it, could you give me a hint as to how I might implement this with co?

**Comment by madbence** [comment]
If you are using thunks, you can use `co-parallel`. If you are using promises, you should roll out your own scheduling algorithm:

```js
co(function* () {
  var urls = [...], done = [];
  while(urls.length) {
    done = done.concat(yield urls.slice(0, n).map(proc)); // take the first n urls and proc them
    urls = urls.slice(n); // drop the proc'd urls
  }
}); 
```

**Comment by targos** [comment]
Use `urls.splice(0, n)` and you don't need line 5.
This is a simple solution but it is not ideal : let's say you want max 5 requests at the same time and 4 of them take 100ms but the last one 1s for some reason. For most of the time you will have only one concurrent request.

It would be nice to have a library similar to caolan/async, designed for promises.

**Comment by dead-horse** [comment]
https://github.com/tj/co-parallel
https://github.com/cojs/gather
